infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,281 @@
+                sessionToDelayedReplicationEventIdMap.putIfAbsent(session, new LinkedList<>());
+                // makes the value part of the map thread safe.
+                sessionToDelayedReplicationEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                    eventList.add(((LogReplicationEvent) event).getEventId());
+                    return eventList;
+                });
+            } else {
+                sessionToReplicationEventIdMap.putIfAbsent(session, new LinkedList<>());
+                // makes the value part of the map thread safe.
+                sessionToReplicationEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                    eventList.add(((LogReplicationEvent) event).getEventId());
+                    return eventList;
+                });
+            }
+        } else if (fsm.equals(FsmEventType.LogReplicationSinkEvent)){
+            LogReplicationSession session = ((LogReplicationSinkEvent) event).getSourceLeadershipManager().getSession();
+            sessionToSinkEventIdMap.putIfAbsent(session, new LinkedList<>());
+            // makes the value part of the map thread safe.
+            sessionToSinkEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                eventList.add(((LogReplicationSinkEvent) event).getEventId());
+                return eventList;
+            });
+        } else {
+            log.warn("Ignored {}. Unexpected FSM event type", fsm);
+        }
+    }
+
+    private void processSinkTask(LogReplicationSinkEvent event) {
+        LogReplicationSession session = event.getSourceLeadershipManager().getSession();
+        synchronized (event.getSourceLeadershipManager()) {
+            while (!sessionToSinkEventIdMap.get(session).get(0).equals(event.getEventId())) {
+                try {
+                    event.getSourceLeadershipManager().wait();
+                } catch (InterruptedException e) {
+                    log.error("Wait for session {} was interrupted {}", session, e.getMessage());
+                }
+            }
+        }
+
+        event.getSourceLeadershipManager().processEvent(event);
+
+        synchronized (event.getSourceLeadershipManager()) {
+            sessionToSinkEventIdMap.get(session).remove(0);
+            event.getSourceLeadershipManager().notifyAll();
+        }
+    }
+
+    private void processRuntimeTask(LogReplicationRuntimeEvent event) {
+        LogReplicationSession session = event.getRuntimeFsm().getSession();
+        // for a given session, The fsm event should be processed in the order they are queued. This condition ensures
+        // that in the event of 2 threads contending for the monitor, only the task submitted first would be processed.
+        synchronized (event.getRuntimeFsm()) {
+            while (!sessionToRuntimeEventIdMap.get(session).get(0).equals(event.getEventId())) {
+                try {
+                    event.getRuntimeFsm().wait();
```

Comment:
No, this is not possible because we synchronize on the session's runtimeFSM for both wait and notify.
Only the actual processing is not protected by a monitor because its guaranteed to be executed by a single thread only...